libqtile/widget/pulse_volume.py

```diff
@@ -0,0 +1,253 @@
+            self.handle
+        )
+        self._subscribed = True
+
+    def on_sink_update(self, event_type, sink_index):
+        self.timeout_add(0.1, self.get_sinks)
+
+    def wait_for_operation(self, op, unref=True):
+        state = lib.pa_operation_get_state(op)
+        while state == lib.PA_OPERATION_RUNNING:
+            lib.pa_mainloop_iterate(self.loop, 0, ffi.NULL)
+            state = lib.pa_operation_get_state(op)
+
+    def change_volume(self, volume):
+        """
+        order pulseaudio to apply new volume
+        """
+        # store new volume to "speed up" widget update so that we don't have
+        # to wait a callback from pulseaudio
+        self.default_sink['values'] = list(volume.values)
+        op = lib.pa_context_set_sink_volume_by_index(
+            self.context,
+            self.default_sink['index'],
+            volume,
+            ffi.NULL,
+            ffi.NULL
+        )
+        if op:
+            self.wait_for_operation(op)
+
+    def mute_volume(self):
+        op = lib.pa_context_set_sink_mute_by_index(
+            self.context,
+            self.default_sink['index'],
+            not self.default_sink['muted'],
+            ffi.NULL,
+            ffi.NULL
+        )
+        self.wait_for_operation(op)
+
+    def increase_volume(self, value=2):
+        base = self.default_sink['base_volume']
+        volume = ffi.new('pa_cvolume *', {
+            'channels': self.default_sink['channels'],
+            'values': self.default_sink['values'],
+        })
+        lib.pa_cvolume_inc(
+            volume,
+            int(value * base / 100),
+        )
+        # check that we dont go over 100% in case its set in config
+        if self.limit_max_volume:
+            volume.values = [(i if i <= base else base) for i in volume.values]
+        self.change_volume(volume)
+
+    def decrease_volume(self, value=2):
+        volume_level = int(value * self.default_sink['base_volume'] / 100)
+        if not volume_level and max(self.default_sink['values']) == 0:
+            # can't be lower than zero
+            return
+        volume = ffi.new('pa_cvolume *', {
+            'channels': self.default_sink['channels'],
+            'values': self.default_sink['values'],
+        })
+        lib.pa_cvolume_dec(volume, volume_level)
+        self.change_volume(volume)
+
+    def button_press(self, x, y, button):
+        if self.default_sink and button == BUTTON_DOWN:
+            self.decrease_volume(self.step)
+        elif self.default_sink and button == BUTTON_UP:
+            self.increase_volume(self.step)
+        elif self.default_sink and button == BUTTON_MUTE:
+            self.mute_volume()
+        elif button == BUTTON_RIGHT:
+            if self.volume_app is not None:
+                subprocess.Popen(self.volume_app)
+
+        self.tick()
+
+    def poll(self):
+        lib.pa_mainloop_iterate(self.loop, 0, ffi.NULL)
+
+    def tick(self):
+        self.poll()
+
+    def update(self):
```

Comment:
Seems like this is overriding the InLoopPollText update() method?

Reply:
it is overriding original's widget update as it states in the doc since that widget assumes that the value is stored in self.volume so it seems like a best place to update it

Reply:
I think you just don't want to inherit from InLoopPollText() at all, and manage the event loop on your own (or really, subscribe to the main eventloop somehow).